### PR TITLE
[Xedra Evolved] *something smart about dreams*

### PR DESCRIPTION
**Pull request by GuardianDll**
#### Summary
Mods "[Xedra Evolved] Bunch of stuff for dreamer class"
#### Describe the solution
Spells:
Unfold - x rays from your eyes
Summon Shifter - just a melee summon with cool aestetic
Hammering - temporal shaping a pretty nice tool/weapon to use
Summon Duplicator - summon a monster, that will attack another monsters by summoning his own copies
Toss a Coin - short range teleport
Summon Angel - good melee help, but fragile against crowds
Summon Light - heavy artillery, it will destroy every enemy on his pach, but also everything else. Summon it, and run as fast as you can
Construct a Winch - hold you current coordinates, and allow you to turn back to this coordinates afterwards

Fixes:
Updated the eat dream lump spels to use correct mana recovery, equal to amount of dreamdross. eat ingot was tricky, because the amount of mana is just too big, so now it gives pretty nice buff
Fixed helmet_inventor was unable to recharge, because i made a stupid mistake and didn't spot it before
Fixed mirror_artifact to give correct spell
Increased the probabillity to find anything it junk loot groups, because i was stupid and accidentally made a chance to find anything lesser than 0.1%
Made a sledge hammer for a bit less useful as a rifle, and more as a melee - throwing damage is decreased, same as melee, but added an enchantment, that increase his damage each spell level
#### Testing
Everything works pretty good, except angel do not cast a spell after death, for some reason

### CODE REVIEW

File: data/mods/Xedra_Evolved/monsters/monster.json [hunk 1]
```diff
@@ -32,5 +32,183 @@
...
+      }
+    ],
+    "death_function": { "message": "The %s disappears.", "corpse_type": "NO_CORPSE" },
+    "flags": [ "SEES", "HEARS", "SMELLS", "NO_BREATHE" ]
+  },
+  {
+    "id": "mon_duplicator",
+    "type": "MONSTER",
+    "name": { "str": "duplicator" },
+    "description": "Humanoid on two legs, this creature poke on everything around with all his eight eyes, arranged in a circle .  Time to time it split in two, where the one is left on a guard, and another attack the enemy with his long fists.",
```
Comment (NetSysFire) [inline data/mods/Xedra_Evolved/monsters/monster.json:82]:
```suggestion
    "description": "A humanoid on two legs, this creature spies on everything around with its eight eyes, arranged in a circle.  From time to time it splits in two, where the one on the left is in a defensive position, and the other attacks the enemy with its long fists.",
```

File: data/mods/Xedra_Evolved/monsters/monster.json [hunk 2]
```diff
@@ -32,5 +32,183 @@
...
+      }
+    ],
+    "death_function": { "message": "The %s disappears.", "corpse_type": "NO_CORPSE" },
+    "flags": [ "SEES", "HEARS", "SMELLS", "NO_BREATHE", "KEEP_DISTANCE", "PACIFIST" ]
+  },
+  {
+    "id": "mon_duplicator_pawn",
+    "type": "MONSTER",
+    "name": { "str": "duplicator" },
+    "description": "Humanoid on two legs, this creature poke on everything around with all his eight eyes, arranged in a circle .  Time to time it split in two, where the one is left on a guard, and another attack the enemy with his long fists.",
```
Comment (NetSysFire) [inline data/mods/Xedra_Evolved/monsters/monster.json:121]:
```suggestion
    "description": "A humanoid on two legs, this creature spies on everything around with its eight eyes, arranged in a circle.  From time to time it splits in two, where the one on the left is in a defensive position, and the other attacks the enemy with its long fists.",
```

File: data/mods/Xedra_Evolved/monsters/monster.json [hunk 3]
```diff
@@ -32,5 +32,183 @@
...
+    },
+    "regenerates": 25,
+    "regen_morale": true,
+    "flags": [ "SEES", "SMELLS", "SWIMS", "PLASTIC", "FLIES", "ACIDPROOF", "NOHEAD", "NOGIB" ]
+  },
+  {
+    "id": "mon_sapient_light",
+    "type": "MONSTER",
+    "name": { "str": "light" },
+    "description": "Flying piece of light in a shape of… cube?  It constantly change his form, and each his attack tear the piece of space from the reality, making a ridiculously loud explosion.  Even if it's friendly, some of his action are impossible to explain.",
```
Comment (NetSysFire) [inline data/mods/Xedra_Evolved/monsters/monster.json:188]:
```suggestion
    "description": "Flying piece of light in a shape of… a cube?  It constantly changes its form, and each of its attacks tears the piece of space from the reality, causing a ridiculously loud explosion.  Even if it's friendly, some of its action are impossible to explain.",
```

File: data/mods/Xedra_Evolved/monsters/monsterattacks.json [hunk 4]
```diff
@@ -113,5 +113,99 @@
...
+    "valid_targets": [ "self", "ground", "none" ],
+    "flags": [ "SILENT", "RANDOM_TARGET" ],
+    "min_aoe": 4,
+    "max_aoe": 4
+  },
+  {
+    "id": "duplicate_shifter",
+    "type": "SPELL",
+    "name": "Summon Shifter (Monster Version)",
+    "description": "Summon a shifter.  Monster version, the creature exist 1-2 seconds.  Mostly decorative effect to show 'glitchness'.",
```
Comment (NetSysFire) [inline data/mods/Xedra_Evolved/monsters/monsterattacks.json:133]:
```suggestion
    "description": "Summon a shifter.  Monster version, the creature exists for 1-2 seconds.  Mostly decorative effect to show 'glitchiness'.",
```

File: data/mods/Xedra_Evolved/spells/dreamer_spells.json [hunk 5]
```diff
@@ -164,5 +164,219 @@
...
+    "energy_increment": -10,
+    "field_id": "fd_clairvoyant",
+    "min_field_intensity": 1,
+    "max_field_intensity": 1
+  },
+  {
+    "id": "summon_shifter",
+    "type": "SPELL",
+    "name": "Summon Shifter",
+    "description": "Summon portal creature to help in danger situation.",
```
Comment (NetSysFire) [inline data/mods/Xedra_Evolved/spells/dreamer_spells.json:201]:
```suggestion
    "description": "Summon a portal creature to help in a dangerous situation.",
```

File: data/mods/Xedra_Evolved/spells/dreamer_spells.json [hunk 6]
```diff
@@ -164,5 +164,219 @@
...
+    "flags": [ "VERBAL", "CONCENTRATE", "NO_EXPLOSION_SFX" ],
+    "min_duration": 36000,
+    "max_duration": 1080000,
+    "duration_increment": 36000
+  },
+  {
+    "id": "summon_sapient_light",
+    "type": "SPELL",
+    "name": "Summon Light",
+    "description": "Summon a disturbed sapient piece of another reality, that like to destroy living creatures.  It won't attack you.  At least now.",
```
Comment (NetSysFire) [inline data/mods/Xedra_Evolved/spells/dreamer_spells.json:228]:
```suggestion
    "description": "Summon a disturbed sapient entity of another reality, that likes to destroy living creatures.  It won't attack you.  At least for now.",
```

File: data/mods/Xedra_Evolved/spells/dreamer_spells.json [hunk 7]
```diff
@@ -164,5 +164,219 @@
...
+    "min_damage": 360,
+    "max_damage": 500,
+    "min_range": 1,
+    "max_range": 1
+  },
+  {
+    "id": "summon_duplicator",
+    "type": "SPELL",
+    "name": "Summon Duplicator",
+    "description": "Summon portal creature to help in danger situation.",
```
Comment (NetSysFire) [inline data/mods/Xedra_Evolved/spells/dreamer_spells.json:311]:
```suggestion
    "description": "Summon a portal creature to help in a dangerous situation.",
```

File: data/mods/Xedra_Evolved/items/inventor/artifacts.json [hunk 8]
```diff
@@ -102,8 +102,8 @@
     "type": "TOOL",
     "id": "mace_magic",
     "name": { "str": "wolf sledge" },
-    "description": "A long, solid, crafted steel haft, and a massive, comma-like head.  It feels really massive, but it is easy to wield.  You even can whisper ancient words to it while you throw it, and it will always return to your hands.",
-    "weight": "2950 g",
+    "description": "A long, solid, crafted steel haft, and a massive, comma-like head.  It feels really massive, but it is easy to wield, and each your level it deal more and more damage.  You even can whisper ancient words to it while you throw it, and it will always return to your hands.",
```
Comment (NetSysFire) [inline data/mods/Xedra_Evolved/items/inventor/artifacts.json:105]:
```suggestion
    "description": "A long, solid, crafted steel haft, and a bulky, comma-like head.  It feels really massive, but it is surprisingly easy to wield, and you will do more damage as your level increases.  You can make it return by speaking its command word, which happens to be in an ancient, long forgotten language, while throwing.",
```

Comment (GuardianDll) [reply data/mods/Xedra_Evolved/items/inventor/artifacts.json:105]: why it must be `in an ancient, long forgotten language`?

Comment (NetSysFire) [reply data/mods/Xedra_Evolved/items/inventor/artifacts.json:105]: i dunno. i just reworded it so it sounds better. if it doesnt fit, remove the long forgotten part.

File: data/mods/Xedra_Evolved/items/ethereal_items.json [hunk 9]
```diff
@@ -141,5 +141,22 @@
         "move_cost": 100
       }
     ]
+  },
+  {
+    "type": "GENERIC",
+    "id": "constructed_hammer",
+    "name": { "str": "constructed hammer" },
+    "description": "A long handle end with two side head, one of which is hollow, and another… do not exist.  Weirdly, both heads are equally useful, and the balance of hammer is pretty good to be useful as a weapon.  By activating and attacking with the hollow part, the hammer creates a tremendous force, able to break a thick tree (or someone's spine) in half, but will be destroyed afterwards.",
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/ethereal_items.json:149]:
```suggestion
    "description": "A long handle end with two side head, one of which is hollow, and another… which doesn't exist.  Weirdly, both heads are equally useful, and the balance of hammer is pretty good to be useful as a weapon.  By activating and attacking with the hollow part, the hammer creates a tremendous force, able to break a thick tree (or someone's spine) in half, but will be destroyed afterwards.",
```

File: data/mods/Xedra_Evolved/items/spell_learning_items/spell_learning_items_dreamer.json [hunk 10]
```diff
@@ -62,6 +62,119 @@
...
-    "use_action": { "type": "learn_spell", "spells": [ "spell_speed_wonder" ] }
+    "use_action": { "type": "learn_spell", "spells": [ "spell_summon_twin" ] }
+  },
+  {
+    "id": "artifact_dreamer_clairvoyance",
+    "looks_like": "glasses_eye",
+    "type": "ARMOR",
+    "category": "artifacts",
+    "name": { "str": "pair of glasses", "str_pl": "pairs of glasses" },
+    "description": "Pair of a common glasses.  They have no lens or glass inside, only casing.",
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/spell_learning_items/spell_learning_items_dreamer.json:73]:
```suggestion
    "description": "Pair of a common glasses.  They have no lens or glass inside, only casing.  But there is a strange engraving of vines, snakes, and foxes covering every surface of the frames.  You think that maybe the engravings are moving when you aren't looking at them.",
```

File: data/mods/Xedra_Evolved/items/spell_learning_items/spell_learning_items_dreamer.json [hunk 11]
```diff
@@ -62,6 +62,119 @@
...
+    "environmental_protection": 1,
+    "armor": [ { "coverage": 75, "covers": [ "eyes" ], "rigid_layer_only": true } ],
+    "use_action": { "type": "learn_spell", "spells": [ "spell_dreamer_clairvoyance" ] }
+  },
+  {
+    "id": "artifact_shifter",
+    "type": "GENERIC",
+    "category": "artifacts",
+    "name": { "str": "sphere" },
+    "description": "Small, black sphere.  It constantly changes its position.",
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/spell_learning_items/spell_learning_items_dreamer.json:91]:
```suggestion
    "description": "Small, black sphere.  It constantly changes its position.  It even tries to roll out of positions you wedge it into.",
```

File: data/mods/Xedra_Evolved/items/spell_learning_items/spell_learning_items_dreamer.json [hunk 12]
```diff
@@ -62,6 +62,119 @@
...
+    "symbol": "0",
+    "color": "black",
+    "use_action": { "type": "learn_spell", "spells": [ "summon_shifter" ] }
+  },
+  {
+    "id": "artifact_sapient_light",
+    "type": "GENERIC",
+    "category": "artifacts",
+    "name": { "str": "flashlight" },
+    "description": "This is a typical household flashlight.  It has no battery holder.",
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/spell_learning_items/spell_learning_items_dreamer.json:107]:
```suggestion
    "description": "This looks like a typical household flashlight.  It has no battery holder and it hums when you touch it.",
```

File: data/mods/Xedra_Evolved/items/spell_learning_items/spell_learning_items_dreamer.json [hunk 13]
```diff
@@ -62,6 +62,119 @@
...
+    "color": "black",
+    "use_action": { "type": "learn_spell", "spells": [ "summon_sapient_light" ] }
+  },
+  {
+    "id": "artifact_constructed_hammer",
+    "type": "GENERIC",
+    "category": "artifacts",
+    "name": { "str": "nail" },
+    "looks_like": "nail",
+    "description": "A pin-shaped piece of iron.  You feel you need to hammer it.",
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/spell_learning_items/spell_learning_items_dreamer.json:124]:
```suggestion
    "description": "A pin-shaped piece of iron.  You feel the need to hammer it.",
```

**Review by Maleclypse — COMMENTED** [review]
Thank you! I'll try and finish reviewing it soon.

File: data/mods/Xedra_Evolved/items/ethereal_items.json (hunk 9 above)
Comment (GuardianDll) [reply data/mods/Xedra_Evolved/items/ethereal_items.json:149]: The idea is here that second part do not exist - there is just nothing, you can move your hand freely - yey when you attack with it, it works like there is something.

Comment (Maleclypse) [reply data/mods/Xedra_Evolved/items/ethereal_items.json:149]: Yep, I love that idea.  I was trying to make it slightly clearer but I may have failed at that.

File: data/mods/Xedra_Evolved/items/inventor/artifacts.json [hunk 14]
```diff
@@ -189,13 +197,17 @@
...
     "pocket_data": [
       {
-        "pocket_type": "MAGAZINE",
+        "pocket_type": "MAGAZINE_WELL",
+        "rigid": true,
         "airtight": true,
         "watertight": true,
-        "rigid": true,
-        "ammo_restriction": { "battery": 5000 }
+        "flag_restriction": [ "BATTERY_LIGHT", "BATTERY_MEDIUM", "BATTERY_HEAVY" ],
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/items/inventor/artifacts.json:207]:
```suggestion
        "flag_restriction": [ "BATTERY_LIGHT", "BATTERY_MEDIUM" ],
```
Isn't the heavy a multi pound battery?

File: data/mods/Xedra_Evolved/monsters/monster.json [hunk 15]
```diff
@@ -32,5 +32,183 @@
     "harvest": "arachnid",
     "fungalize_into": "mon_spider_fungus",
     "flags": [ "SEES", "SMELLS", "HEARS", "WEBWALK", "CLIMBS", "PATH_AVOID_FIRE", "PATH_AVOID_FALL" ]
+  },
+  {
+    "id": "mon_shifter",
+    "type": "MONSTER",
+    "name": { "str": "shifter" },
+    "description": "Tall body on a four legs, surrounded with a \"glitched\" aura - it constantly changes its own position.",
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/monsters/monster.json:40]:
```suggestion
    "description": "Tall body on four legs, surrounded with a \"glitched\" aura - it constantly changes its own position.",
```

**Review by Maleclypse — APPROVED** [review]
Thank you for this!

File: data/mods/Xedra_Evolved/items/inventor/artifacts.json (hunk 14 above)
Comment (Maleclypse) [reply data/mods/Xedra_Evolved/items/inventor/artifacts.json:207]: If you feel like you've balanced this around the heavy battery it'll be fine.

Comment (GuardianDll) [reply data/mods/Xedra_Evolved/items/inventor/artifacts.json:207]: I think make the balance around middle one is okay

**Pull request merged**